serialized_data_interface/__init__.py

```diff
@@ -148,17 +148,26 @@ def _relations(self):
             rel for rel in self.charm.model.relations[self.relation_name] if rel.app
         ]
 
-    def get_data(self) -> Dict[Tuple[Relation, Application], Dict]:
+    def get_data(
+        self, include_self: bool = False
```

Comment:
I'm not a fan of the flag. We definitely need a way to read the data we have sent over the relation in addition to the data received from the other side, but when you need the sent data you certainly don't want to troll through the received data as well to get it, just like you don't want to troll through the received data to manually exclude the sent data.

If I were designing this from scratch, I would make it mirror the framework model much more closely. Specifically, I would require that you pass in the relation and app or unit so that it's clear and explicit what data you want. You generally need that info anyway, as evinced by the fact that you're returning those as the keys to the dict, so it doesn't seem any less clean to just iterate over them explicitly from the outside in rather than having them collected from the inside out, and it also makes it clear when say `get_data(rel, self.app)` that you're talking about your own data rather than remote data (`get_data(rel, rel.app)` although I do kind of which that were named `rel.remote_app` instead). It also gives you a much better place to raise errors that are specific to a single relation, unit, or app, without adversely impacting others, and makes it easier to solve #27, #28, and #31.

Reply:
@johnsca I tried adding in support for `get_data(rel, self.model.app)`. It works, but if you've got both `rel` and `app`, why not just use `rel.data[app]` anyways and skip `get_data()` entirely? That allows for any specific control you might need for handling errors, allows you to optionally skip schema validation, and other sorts of things like that.

Updated this PR to drop the flag and just exclude self app unconditionally.

Reply:
`get_data` does more than just collect the data from the various relations (also, FWIW, having `rel` implies having `app`, since it's either `rel.app` or your local app); it also handles:

* The implicit `data` key that everything is shoved into
* Deserialization of the data from YAML text
* Determining the appropriate schema based on the common support version
* Validating the deserialized data against the appropriate schema

I'm totally in favor factoring that logic out into something like a `sdi.get(rel, app_or_unit)` and renaming `sdi.get_data()` to `sdi.get_all()` as a helper for loading everything (even though I personally would never use it, because it doesn't give enough control over issues with specific relations, apps, or units, as I mentioned previously).  I'd also like to see the `sdi._relations` property be made public to support that usage pattern.

Reply:
Also not opposed to some more descriptive name that indicates that it is doing those extra things, but I also recognize that it will be used frequently, so a shorter name is likely preferable even if less descriptive.

Reply:
I don't think that iterating over the relations inherently make it more bug prone as long as SDI is still the thing doing the deserialization, and it's closer to the way the Operator Framework presents the relations in the first place, so it will be easier to learn.  I also like the name `unwrap` a lot better than `get_data` or `get`, since it really conveys that it's doing something other than just pulling the data out as-is while still being concise. Finally, instead of a flag to exclude the local app, we could use the same style of data structure as the framework with the apps / units as the keys. So, combining all of that, what do you think of something like:

```python
    def _on_changed(self, event):
        interface = get_interface(self, "endpoint")
        for rel in self.model.relations["endpoint"]:
            try:
                data = interface.unwrap(rel)
            except SDIError as e:
                log.error(f"Relation failed: {e}")
                continue
            local_data = data[self.app]
            remote_data = data[rel.app]
            local_data["field"] = "foo"
            interface.wrap(rel, local_data)
```

Reply:
Actually, since `event.relation` is a thing, you could even tweak the example to avoid the loop entirely and just process the one relation that changed. That would be useful in, e.g., the `istio-pilot` charm to avoid recreating `VirtualService` and `Gateway` entries for relations which haven't changed.

Reply:
Doing just `log.error(f"Relation failed: {e}")` makes it very easy to not notice that anything is wrong. I would qualify that as a bug, and what I mean by bug-prone. `juju status` will show everything as happily humming along, and someone will have to know to go check the logs for the appropriate charm. You'll want to set `BlockedStatus` for the charm, and at that point, why not just use `get_data()` as-is? That will ensure you never forget to properly handle error cases.

And I don't think it's a bad idea to iterate over the relations yourself if you want to skip that built-in error checking and do fine-grained handling. Something like this:

```python
    def _on_changed(self, event):
        interface = get_interface(self, "endpoint")
        for rel in self.model.relations["endpoint"]:
            schema = interface.schema[interface.versions[rel.app.name]]
            data = yaml.safe_load(rel.data['data'])
            validate(data, schema)
            # Do whatever with `data`
            validate(data, schema)
            rel.data[self.app]['data'] = yaml.safe_dump(data)
```

Sprinkle exception handling to taste. As a side note, I almost forgot to include the second `validate(data, schema)` line, which is also a bug, and also why I think the existing SDI functions should be used as-is.

Reply:
RE: using `event.relation` and skipping the loop, I'd rather just be able to inform Juju of what exactly what YAML should exist, and let it do the work of actively keeping that YAML in line. Using a loop in `_on_changed` acts as a poor man's version of that, ensuring that all `VirtualService`s haven't suffered from configuration drift when any of them changes. The delete/create cycle also removes old `VirtualService`s that shouldn't exist any more, unlike a regular `kubectl apply`.

Reply:
Your example is the same as mine except that I want to keep the schema, deserialize, validate logic encapsulated in the library via the `.unwrap()` and `.wrap()` to avoid the bugs you mentioned and to keep the implementation details internal. I don't care as much if the interface is `interface.unwrap(rel)` or `result.unwrap()` since the usage is almost identical, but there's also no reason the existing interface couldn't be kept as a helper around the more explicit per-rel behavior to enable both patterns.

Reply:
> I'd rather just be able to inform Juju of what exactly what YAML should exist, and let it do the work of actively keeping that YAML in line.
> ...
> The delete/create cycle also removes old VirtualServices that shouldn't exist any more, unlike a regular kubectl apply.

Whether it's Juju having apply-like semantics or `kubectl` / lightkube, I don't see how either one of them could handle removing things that aren't in the new manifest but were in the old. And you won't see any changes to relation data outside of a `relation-changed` hook, so I'm not sure what you mean by "configuration drift". Also, `relation-changed` isn't the right place to remove things if you're doing it incrementally; that would be `relation-broken` (or maybe `relation-departed` if it's per-unit).

Either way, the only real benefit of moving away from the full delete + reapply is that it would prevent network disruption to services which didn't change just because a new service was added.  I'm not sure how big of an issue that is, but I'm fine waiting to fix it when / if it does become a problem.

Reply:
(I'm crafting an on-topic reply, but separately want to say: ) I do worry about network disruptions with the "delete all, recreate all" strategy.  Not because I know they'll definitely be a problem (I have no idea), but because I'm confident if they are a problem, they'll be a nightmare to debug

Reply:
I think we've got a few conversations on the go now so I am only responding to a subset...

It feels like there's agreement that the current `get_data()` is broad, broadness can lead to problems, and that at best it only needs to be this broad in a small subset of use cases.  When someone needs data via SDI they often need only a subset, and if the interface let them ask for that it would be a nice user experience.  

I think I'm with @johnsca in that it would be nice to have the api let me do things like:
```python
data = interface.some_data_getting_function(rel)
local_data = data[self.app]
remote_data = data[rel.app]
```
with SDI I think we are trying to make relation data easier to interact with and require less code to do common things, and something like this achieves both.

Personally I might take it a step further and overload the `some_data_getting_function` so that I could optionally provide `rel` and `app` and it returns the relevant info.  To me it feels having the api a bit more declarative might be easier to explain to general users.  In general as is I think right now we ask general users to understand the guts too much, and that we want to move away from that. 

@knkski I'm not against someone iterating on the relations by themselves if they've got a really good reason, but I don't think there will be many good reasons for that in general use.  I don't think many people could correctly come up with the code block you propose, and at that point why do we have an SDI at all?

Reply:
I am with you @knkski in the worry about configuration drift, but I don't think the package that serialises/deserialises relation data is the intuitive place to combat this.  Except in places where we make meaningful simplifications to hide detail, keeping the data structures close to how the framework has them sounds like a good idea that'll help people use this tool, and also in this case it simplifies things.